### PR TITLE
Add benchmark:{run, diff} mage targets

### PR DESCRIPTION
**Pull request by alexsapran**
This PR adds 2 new mage targets for the project that execute the go bench benchmarks and benchstat for comparing the results.

Example use of the above is:
```
$ mage benchmark:run benchmark-before
# ..... make changes ......
$ mage benchmark:run benchmark-after
$ mage benchmark:diff benchmark-before benchmark-after
```

Once we are happy with the workflow we can move this part into the `elastic-agent-libs/dev-tools/mage` and reference from there.

The intention of these new mage targets is to be used when adding code to this repo with the above workflow and to integrate them into a GH action and have it report the numbers back to the PR, an example workflow that I was playing around with can be found in https://github.com/alexsapran/elastic-agent-shipper-client/pull/1#issuecomment-1450737792

### CODE REVIEW

**Comment by faec** [comment]
Sorry, I'm lacking some context -- this repository is for boilerplate API definitions used to make RPC calls to the shipper, it doesn't run anything by itself. What's being benchmarked?

**Comment by alexsapran** [comment]
Hi @faec, you were quicker to comment than me writing all the context. 

We introduced micro benchmarks in https://github.com/elastic/elastic-agent-shipper-client/pull/17 and intended to have them as part of the workflow of future changes. 
I will be pushing another commit as well to add the GH action that would use the new mage targets. I am asking you to hold reviewing until I am finished. You were automatically got assigned while opening the PR :D

**Comment by alexsapran** [comment]
I think I am done with the changes in this PR, I would have to do a follow-up that only adds the gh action that is going to be using those targets.

File: magefile.go [hunk 1]
```diff
@@ -181,3 +196,147 @@ func licenser(mode licenserMode) error {
...
+		args = append(args, filepath.Join("github.com/elastic/elastic-agent-shipper-client", pkg, "..."))
+	}
+	goTestBench := makeCommand(ctx, nil, "go", args...)
+
+	// Wire up the outputs.
+	var outputs []io.Writer
+	if outputFile != "" {
+		fileOutput, err := os.Create(createDir(outputFile))
+		if err != nil {
+			return errors.Wrap(err, "failed to create go test output file")
```
Comment (cmacknz) [inline magefile.go]: You want to use the standard library's `errors` package with the `%w` wrapping verb instead:

```suggestion
			return fmt.Errorf("failed to create go test output file: %w", err)
```

We have a lint warning in the other repositories that suggest it but we must be missing it here. The pkg/errors version is not actively maintained anymore and isn't exactly compatible with the standard library version.

File: magefile.go [hunk 2]
```diff
@@ -8,15 +8,22 @@
...
+	"os/exec"
 	"path"
+	"path/filepath"
+	"strings"
 
 	devtools "github.com/elastic/elastic-agent-libs/dev-tools/mage"
 	"github.com/elastic/elastic-agent-libs/dev-tools/mage/gotool"
 	"github.com/magefile/mage/mg"
 	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
```
Comment (cmacknz) [inline magefile.go]: Replace all uses of this with the stdlib's `errors` package

**Comment by cmacknz** [comment]
We implement common mage targets in https://github.com/elastic/elastic-agent-libs/tree/main/dev-tools/mage if you end up moving these to another repository afterwards and don't want to just copy them.

File: magefile.go [hunk 3]
```diff
@@ -181,3 +196,147 @@ func licenser(mode licenserMode) error {
...
+		outputs = append(outputs, fileOutput)
+	}
+	output := io.MultiWriter(outputs...)
+	goTestBench.Stdout = io.MultiWriter(output, os.Stdout)
+	goTestBench.Stderr = io.MultiWriter(output, os.Stderr)
+
+	err := goTestBench.Run()
+
+	var goTestErr *exec.ExitError
+	if err != nil {
```
Comment (fearful-symmetry) [inline magefile.go:244]: I'm not sure why the error checks are in two separate if statements, it's a tad hard to follow at first. It might be a little more concise if we just do something like `if errors.Is() {...} else if errors.Is() {...} else err != nil {...}` or something like that.

File: magefile.go [hunk 4]
```diff
@@ -181,3 +196,147 @@ func licenser(mode licenserMode) error {
...
+
+func InstallBenchStat() error {
+	err := gotool.Install(gotool.Install.Package(goBenchstats))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (Benchmark) Run(ctx context.Context, outputFile string) error {
```
Comment (fearful-symmetry) [inline magefile.go]: I wonder if there's a way to make this a tad more user friendly, since if you omit the output file, you get a rather unhelpful `not enough arguments for target "Benchmark:Run", expected 1, got 0` error message from mage. Just adding a godoc that says something like `// Run all benchmarks. Takes the name of an output file as an argument` might help. Ditto with `benchmark:diff`

File: magefile.go (hunk 1 above)
Comment (alexsapran) [reply magefile.go]: Thanks for the info, I was not aware of this. I used a pre-existing code from `beats`, I am guessing that lint has a bug?

File: magefile.go (hunk 3 above)
Comment (alexsapran) [reply magefile.go:244]: I had not put too much thought into this part of the build as my main focus was the github action.
It is a copy/paste from `beats` https://github.com/elastic/beats/blob/c3f5d7eed9405f71fea618ffa53e895c3c908aff/dev-tools/mage/gotest.go#L271-L278

I will try your suggestion 👍🏼

File: magefile.go (hunk 1 above)
Comment (cmacknz) [reply magefile.go]: Not a bug in the linter, the Beats code was just written before the standard library included the features from pkg/errors and rewriting all of it will take a long time.

File: magefile.go (hunk 4 above)
Comment (alexsapran) [reply magefile.go]: please see [66c329c](https://github.com/elastic/elastic-agent-shipper-client/pull/22/commits/66c329c2fd2b0f34ecfd6252edab1906995ddb93) if makes this part any better.

File: magefile.go [hunk 5]
```diff
@@ -181,3 +196,144 @@ func licenser(mode licenserMode) error {
...
+
+	if goTestErr != nil {
+		// No packages were tested. Probably the code didn't compile.
+		return goTestErr
+	}
+
+	return nil
+}
+
+func (Benchmark) Diff(ctx context.Context) error {
```
Comment (fearful-symmetry) [inline magefile.go:262]: Can we add godocs to the new mage targets? The rest of them have docs, and it's nice for usability/consistency: 
```
  benchmark:diff
  benchmark:run
  check                  runs all the checks
  checkLicenseHeaders    checks ASL2 headers in .go files
  generateGo             regenerates the Go files out of .proto files
  installBenchStat
  installLicenser        installs the go-licenser.
  installProtoGo         installs required plugins for protoc
  license                applies the right license header.
  update                 updates all the generated code out of the spec
```

File: magefile.go [hunk 6]
```diff
@@ -181,3 +196,144 @@ func licenser(mode licenserMode) error {
...
+	baseFile := os.Getenv("BASE")
+	nextFile := os.Getenv("NEXT")
+	var args = []string{}
+	if baseFile == "" {
+		log.Printf("Missing baseline benchmark output, benchstat will show a summary for the bechmark")
+	} else {
+		args = append(args, baseFile)
+	}
+	if nextFile == "" {
+		return fmt.Errorf("Missing candidate benchmark output file, please run first OUTPUT=foo benchmark:run")
```
Comment (fearful-symmetry) [inline magefile.go]: Small nit: can you make the `OUTPUT=foo` line a bit more descriptive? Something like `OUTPUT=path_to_other_benchmark_output` or something like that.

**Review by fearful-symmetry — DISMISSED** [review]

**Comment by alexsapran** [comment]
I am going to close as I have started the work to migrate this into [elastic-agent-libs](https://github.com/elastic/elastic-agent-libs/pull/110)
